lib/cli/src/Cardano/CLI.hs

```diff
@@ -409,25 +431,103 @@ hoistKeyScheme eta s = CliKeyScheme
     , inspect = eta . inspect s
     }
 
+-- | Bidirectional encoding of data that allows specialization.
+--
+-- When decoding, the correponding, potentially specialized encoder is returned
+-- along with the result:
+-- >>> let (Codec _ decode) = ...
+-- >>> (x, encodeBack) <- decode
+--
+-- This allows us to have a @Codec@ for /both/ hex and bech32, and makes sure
+-- that we always encode the result back in the same encoding as we first got.
+data Codec m a b = Codec (a -> m (b, Codec m b a)) (b -> m (a, Codec m a b))
```

Comment:
This type seems overly-abstracted to me. It _could_ (maybe) make sense in an abstract library for manipulating codecs and even there, you'd most likely want to pick a concrete type for your target encoding which will end up being `ByteString`. There's no other pragmatic use-case for a Codec.

Plus, it is very unlikely than encoding a value would fail, and in particular, would fail with the same error type. That would be a silly API. 

What about simplifying this as:

Consider instead:

```hs
data Codec (m :: * -> *) = Codec 
    { encode = ByteString -> Text
    , decode = Text -> m ByteString
    }
```

(and even here, it is already _too abstract_ for our need IMO ..) 

And aim for _simple_ code. Category theory is cool and can help coming up with nice abstractions. But remember that we have to maintain this code afterwards and here, you don't need to represent Codec as a utterly abstract concept to get a code that is flexible enough to support multiple types of encoding. There's a balance between flexibility and simplicity. Here, it is clearly too much on the side of flexibility.

Reply:
```haskell
data Codec (m :: * -> *) = Codec 
    { encode :: ByteString -> Text
    , decode :: Text -> m ByteString
    }
```
Wouldn't work if we want `bech32`, `hex`, `any` to all be values of `Codec`. This is because we need to make sure to encode the value back using the same codec as we got it in, as described in the comment.

In that case one should remove even more abstractions ->  #1599.